samples/apps/copilot-chat-app/SKWebApi/Skills/ChatSkill.cs

```diff
@@ -0,0 +1,354 @@
+﻿using System.Globalization;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.AI;
+using Microsoft.SemanticKernel.AI.TextCompletion;
+using Microsoft.SemanticKernel.CoreSkills;
+using Microsoft.SemanticKernel.Memory;
+using Microsoft.SemanticKernel.Orchestration;
+using Microsoft.SemanticKernel.SkillDefinition;
+
+namespace SemanticKernel.Service.Skills;
+
+/// <summary>
+/// ChatSkill offers a more coherent chat experience by using memories
+/// to extract conversation history and user intentions.
+/// </summary>
+public class ChatSkill
+{
+    /// <summary>
+    /// A kernel instance to create a completion function since each invocation
+    /// of the <see cref="ChatAsync"/> function will generate a new prompt dynamically.
+    /// </summary>
+    private readonly IKernel _kernel;
+
+    public ChatSkill(IKernel kernel)
+    {
+        this._kernel = kernel;
+    }
+
+    /// <summary>
+    /// Extract user intent from the conversation history.
+    /// </summary>
+    /// <param name="context">Contains the 'audience' indicating the name of the user.</param>
+    [SKFunction("Extract user intent")]
+    [SKFunctionName("ExtractUserIntent")]
+    [SKFunctionContextParameter(Name = "audience", Description = "The audience the chat bot is interacting with.")]
+    public async Task<string> ExtractUserIntentAsync(SKContext context)
+    {
+        var tokenLimit = SystemPromptDefaults.CompletionTokenLimit;
+        var historyTokenBudget =
+            tokenLimit -
+            SystemPromptDefaults.ResponseTokenLimit -
+            this.EstimateTokenCount(string.Join("\n", new string[]
+                {
+                    SystemPromptDefaults.SystemDescriptionPrompt,
+                    SystemPromptDefaults.SystemIntentPrompt,
+                    SystemPromptDefaults.SystemIntentContinuationPrompt
+                })
+            );
+
+        var intentExtractionVariables = new ContextVariables();
+        intentExtractionVariables.Set("tokenLimit", historyTokenBudget.ToString(new NumberFormatInfo()));
+        intentExtractionVariables.Set("knowledgeCutoff", SystemPromptDefaults.KnowledgeCutoffDate);
+        intentExtractionVariables.Set("audience", context["audience"]);
+
+        var completionFunction = this._kernel.CreateSemanticFunction(
```

Comment:
nit: Any reason to not initialize this once in the constructor?

Reply:
Let's save this change for fit-and-finish. I'll create a tracking task.